nncf/quantization/algorithms/post_training/algorithm.py

```diff
@@ -188,28 +211,207 @@ def apply(
+        for input_data in tqdm(
+            islice(calibration_dataset.get_inference_data(), calibration_dataset_size),
+            total=calibration_dataset_size,
+            desc=f"Collect dataset for {model_cnt} model:",
+        ):
+            data_item = []
+            results = engine.infer(input_data)
+            if (if_submodel_condition and results[if_cond_input_name]) or (
+                not if_submodel_condition and not results[if_cond_input_name]
+            ):
+                for input_name in child_model_input_names:
+                    data_item.append(results[input_name])
+                dataset.append(data_item)
+        nncf_logger.info(f"The final length of a dataset for {model_cnt} model is {len(dataset)}")
+        return Dataset(dataset)
+
+    def _has_if_op(self, nncf_graph: NNCFGraph, if_node_metatype: OperatorMetatype) -> bool:
+        """
+        Returns True if NNCGraph has If node.
+
+        :param nncf_graph: NNCFGraph instance.
+        :param if_node_metatype: backend-specific If node metatype.
+        :return: True if NNCFGraph has If node, else - otherwise.
+        """
+        if nncf_graph.get_nodes_by_metatypes([if_node_metatype]):
+            return True
+        return False
+
+    def _extract_if_submodel(
+        self, model_transformer: ModelTransformer, if_node: NNCFNode, if_submodel_condition: bool
+    ) -> TModel:
+        """
+        Returns if submodel of If node laying on an input port if_submodel_port_id of If node.
+
+        :param model_transformer: ModelTransformer instance.
+        :param if_node: If node.
+        :param if_submodel_condition: If True returns True submodel of If node, otherwise - False submodel.
+        :return: If submodel.
+        """
+        transformation_layout = TransformationLayout()
+        command = self._backend_entity.create_extract_if_subgraph_command(if_node, if_submodel_condition)
+        transformation_layout.register(command)
+        return model_transformer.transform(transformation_layout)
+
+    def _update_if_submodel(
+        self, model_transformer: ModelTransformer, if_node: NNCFNode, if_submodel_condition: bool, submodel: TModel
+    ) -> TModel:
+        """
+        Update submodel of If node.
+
+        :param model_transformer: ModelTransformer instance.
+        :param if_node: If node.
+        :param if_submodel_condition: Condition of If node submodel.
+        :param submodel: New submodel.
+        :return: Updated model with a new submodel.
+        """
+        transformation_layout = TransformationLayout()
+        command = self._backend_entity.create_update_subgraph_command(if_node, if_submodel_condition, submodel)
+        transformation_layout.register(command)
+        return model_transformer.transform(transformation_layout)
+
+    def _add_outputs_before_if_node(
+        self, model_transformer: ModelTransformer, model: TModel, if_node: NNCFNode
+    ) -> TModel:
```

Comment:
Why do we add outputs before the `if` node?

Reply:
We need extra outputs of original (parent) model to collect dataitems for then/else submodels